File: project-analyzer-java/src/main/java/com/analyzer/module/ArchitectureAnalyzer.java
package com.analyzer.module;

import com.analyzer.model.*;
import com.analyzer.model.AnalysisReport.*;
import com.analyzer.plugin.AnalysisModuleInterface;
import com.analyzer.plugin.LanguagePlugin;

import java.io.IOException;
import java.nio.file.*;
import java.util.*;
import java.util.stream.Stream;

public class ArchitectureAnalyzer implements AnalysisModuleInterface {

    private static final Set<String> IGNORED_DIRS = Set.of(
            "node_modules", ".git", "dist", "build", "target",
            ".gradle", "__pycache__", ".venv", "venv", ".idea", ".vscode"
    );

    private record FrameworkDef(String name, String category, List<String> packages) {}

    private static final List<FrameworkDef> KNOWN_FRAMEWORKS = List.of(
            new FrameworkDef("Spring Boot", "web", List.of("spring-boot-starter-web", "spring-boot-starter")),
            new FrameworkDef("Spring MVC", "web", List.of("spring-webmvc")),
            new FrameworkDef("MyBatis", "orm", List.of("mybatis", "mybatis-spring")),
            new FrameworkDef("Hibernate", "orm", List.of("hibernate-core")),
            new FrameworkDef("Spring Data JPA", "orm", List.of("spring-data-jpa", "spring-boot-starter-data-jpa")),
            new FrameworkDef("Spring Security", "security", List.of("spring-security-core", "spring-boot-starter-security")),
            new FrameworkDef("Spring Cloud", "microservice", List.of("spring-cloud-starter")),
            new FrameworkDef("Express", "web", List.of("express")),
            new FrameworkDef("React", "frontend", List.of("react")),
            new FrameworkDef("Vue", "frontend", List.of("vue")),
            new FrameworkDef("Django", "web", List.of("django")),
            new FrameworkDef("Flask", "web", List.of("flask")),
            new FrameworkDef("Gin", "web", List.of("github.com/gin-gonic/gin"))
    );

    private record LayerPattern(String name, String pattern, String regex) {}

    private static final List<LayerPattern> LAYER_PATTERNS = List.of(
            new LayerPattern("Controller/Handler", "controller*|handler*|route*|api*", "(?i)^(controller|handler|route|api)"),
            new LayerPattern("Service", "service*|business*|domain*", "(?i)^(service|business|domain)"),
            new LayerPattern("Repository/Data", "repo*|repository*|dao*|data*|model*|entity*", "(?i)^(repo|repository|dao|data|model|entity)"),
            new LayerPattern("Utility", "util*|helper*|common*|shared*|lib*", "(?i)^(util|helper|common|shared|lib)")
    );

    @Override
    public String getName() { return "architecture"; }

    @Override
    public Object analyze(ProjectProfile profile, List<LanguagePlugin> plugins) {
        List<Dependency> dependencies = collectDependencies(profile, plugins);
        Map<Dependency.DependencyCategory, List<Dependency>> groups = groupDependencies(dependencies);
        List<FrameworkInfo> frameworks = identifyFrameworks(dependencies);
        List<LayerInfo> layers = identifyLayers(profile.getProjectPath());
        MermaidGraph graph = buildModuleDependencyGraph(profile);

        ArchitectureResult result = new ArchitectureResult();
        result.setDependencies(dependencies);
        result.setDependencyGroups(groups);
        result.setFrameworks(frameworks);
        result.setLayers(layers);
        result.setModuleDependencyGraph(graph);
        return result;
    }

    private List<Dependency> collectDependencies(ProjectProfile profile, List<LanguagePlugin> plugins) {
        List<Dependency> all = new ArrayList<>();
        Set<String> seen = new HashSet<>();
        for (LanguagePlugin plugin : plugins) {
            for (Dependency dep : plugin.extractDependencies(Path.of(profile.getProjectPath()))) {
                String key = (dep.getGroup() != null ? dep.getGroup() : "") + ":" + dep.getName() + ":" + dep.getVersion();
                if (seen.add(key)) all.add(dep);
            }
        }
        return all;
    }

    private Map<Dependency.DependencyCategory, List<Dependency>> groupDependencies(List<Dependency> deps) {
        Map<Dependency.DependencyCategory, List<Dependency>> groups = new EnumMap<>(Dependency.DependencyCategory.class);
        for (Dependency.DependencyCategory cat : Dependency.DependencyCategory.values()) {
            groups.put(cat, new ArrayList<>());
        }
        for (Dependency dep : deps) {
            groups.get(dep.getCategory()).add(dep);
        }
        return groups;
    }

    private List<FrameworkInfo> identifyFrameworks(List<Dependency> deps) {
        Map<String, Dependency> depMap = new HashMap<>();
        for (Dependency d : deps) depMap.put(d.getName().toLowerCase(), d);

        List<FrameworkInfo> frameworks = new ArrayList<>();
        for (FrameworkDef fw : KNOWN_FRAMEWORKS) {
            for (String pkg : fw.packages()) {
                Dependency dep = depMap.get(pkg.toLowerCase());
                if (dep != null) {
                    FrameworkInfo fi = new FrameworkInfo();
                    fi.setName(fw.name());
                    fi.setVersion(dep.getVersion());
                    fi.setCategory(fw.category());
                    fi.setEvidence(List.of("dependency: " + dep.getName() + "@" + dep.getVersion()));
                    frameworks.add(fi);
                    break;
                }
            }
        }
        return frameworks;
    }

    private List<LayerInfo> identifyLayers(String projectPath) {
        List<LayerInfo> layers = new ArrayList<>();
        Path root = Path.of(projectPath);
        List<Path> dirsToScan = new ArrayList<>(List.of(root));
        for (String sub : List.of("src", "lib", "app", "pkg", "internal", "src/main/java")) {
            Path p = root.resolve(sub);
            if (Files.isDirectory(p)) dirsToScan.add(p);
        }

        for (LayerPattern lp : LAYER_PATTERNS) {
            List<String> matchedFiles = new ArrayList<>();
            List<String> matchedClasses = new ArrayList<>();
            var regex = java.util.regex.Pattern.compile(lp.regex());
            for (Path dir : dirsToScan) {
                findMatchingDirs(dir, root, regex, matchedFiles, matchedClasses);
            }
            if (!matchedFiles.isEmpty()) {
                LayerInfo li = new LayerInfo();
                li.setName(lp.name());
                li.setPattern(lp.pattern());
                li.setClasses(matchedClasses);
                li.setFiles(matchedFiles);
                layers.add(li);
            }
        }
        return layers;
    }

    private void findMatchingDirs(Path searchDir, Path root, java.util.regex.Pattern regex,
                                  List<String> files, List<String> classes) {
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(searchDir)) {
            for (Path entry : stream) {
                if (!Files.isDirectory(entry)) continue;
                String name = entry.getFileName().toString();
                if (IGNORED_DIRS.contains(name)) continue;
                if (regex.matcher(name).find()) {
                    listFilesRecursive(entry, root, files, classes);
                }
            }
        } catch (IOException ignored) {}
    }

    private void listFilesRecursive(Path dir, Path root, List<String> files, List<String> classes) {
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(dir)) {
            for (Path entry : stream) {
                if (Files.isDirectory(entry)) {
                    if (!IGNORED_DIRS.contains(entry.getFileName().toString())) {
                        listFilesRecursive(entry, root, files, classes);
                    }
                } else if (Files.isRegularFile(entry)) {
                    files.add(root.relativize(entry).toString());
                    String base = entry.getFileName().toString();
                    int dot = base.lastIndexOf('.');
                    if (dot > 0) classes.add(base.substring(0, dot));
                }
            }
        } catch (IOException ignored) {}
    }

    private MermaidGraph buildModuleDependencyGraph(ProjectProfile profile) {
        var modules = profile.getModules();
        if (modules == null || modules.size() < 2) return null;

        List<String> nodes = modules.stream().map(ProjectProfile.SubModule::getName).toList();
        List<MermaidGraph.MermaidEdge> edges = new ArrayList<>();

        StringBuilder sb = new StringBuilder("graph TD\n");
        for (String node : nodes) {
            sb.append("  ").append(sanitize(node)).append("[\"").append(node).append("\"]\n");
        }

        // Detect cross-module deps by checking build files
        for (var mod : modules) {
            Path modPath = Path.of(profile.getProjectPath(), mod.getPath());
            for (var other : modules) {
                if (other.getName().equals(mod.getName())) continue;
                if (checkModuleDependency(modPath, other.getName())) {
                    edges.add(new MermaidGraph.MermaidEdge(mod.getName(), other.getName(), null));
                    sb.append("  ").append(sanitize(mod.getName())).append(" --> ")
                            .append(sanitize(other.getName())).append("\n");
                }
            }
        }

        MermaidGraph graph = new MermaidGraph();
        graph.setSyntax(sb.toString().trim());
        graph.setNodes(nodes);
        graph.setEdges(edges);
        return graph;
    }

    private boolean checkModuleDependency(Path modPath, String otherName) {
        for (String file : List.of("pom.xml", "build.gradle", "package.json")) {
            Path p = modPath.resolve(file);
            if (Files.exists(p)) {
                try {
                    String content = Files.readString(p);
                    if (content.contains(otherName)) return true;
                } catch (IOException ignored) {}
            }
        }
        return false;
    }

    private String sanitize(String name) {
        return name.replaceAll("[^a-zA-Z0-9_]", "_");
    }
}

File: project-analyzer-ts/src/modules/architecture-analyzer.ts
/**
 * ArchitectureAnalyzer — 技术架构分析模块
 *
 * 实现 AnalysisModuleInterface，提供：
 * - 从所有语言插件收集依赖并按类别分组
 * - 识别核心框架和技术
 * - 识别项目分层结构（Controller/Service/Repository/Utility）
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import type { AnalysisModuleInterface, ModuleResult } from '../plugins/analysis-module.js';
import type { LanguagePlugin } from '../plugins/language-plugin.js';
import type {
  ProjectProfile,
  SubModule,
  Dependency,
  DependencyCategory,
  ArchitectureResult,
  LayerInfo,
  FrameworkInfo,
  MermaidGraph,
} from '../models/index.js';

// ─── Framework detection mapping ───

interface FrameworkDef {
  name: string;
  category: string;
  packages: string[];
}

const KNOWN_FRAMEWORKS: FrameworkDef[] = [
  // JS/TS web frameworks
  { name: 'Express', category: 'web', packages: ['express'] },
  { name: 'Koa', category: 'web', packages: ['koa'] },
  { name: 'Fastify', category: 'web', packages: ['fastify'] },
  { name: 'Hapi', category: 'web', packages: ['hapi', '@hapi/hapi'] },
  { name: 'NestJS', category: 'web', packages: ['@nestjs/core', 'nestjs'] },
  { name: 'Next.js', category: 'web', packages: ['next'] },
  { name: 'Nuxt', category: 'web', packages: ['nuxt'] },
  // Frontend frameworks
  { name: 'React', category: 'frontend', packages: ['react'] },
  { name: 'Vue', category: 'frontend', packages: ['vue'] },
  { name: 'Angular', category: 'frontend', packages: ['@angular/core', 'angular'] },
  { name: 'Svelte', category: 'frontend', packages: ['svelte'] },
  // Python web frameworks
  { name: 'Django', category: 'web', packages: ['django'] },
  { name: 'Flask', category: 'web', packages: ['flask'] },
  { name: 'FastAPI', category: 'web', packages: ['fastapi'] },
  // Go web frameworks
  { name: 'Gin', category: 'web', packages: ['github.com/gin-gonic/gin'] },
  { name: 'Echo', category: 'web', packages: ['github.com/labstack/echo'] },
  { name: 'Fiber', category: 'web', packages: ['github.com/gofiber/fiber'] },
  // ORM / Database
  { name: 'TypeORM', category: 'orm', packages: ['typeorm'] },
  { name: 'Prisma', category: 'orm', packages: ['prisma', '@prisma/client'] },
  { name: 'Sequelize', category: 'orm', packages: ['sequelize'] },
  { name: 'Mongoose', category: 'orm', packages: ['mongoose'] },
  { name: 'SQLAlchemy', category: 'orm', packages: ['sqlalchemy'] },
  { name: 'GORM', category: 'orm', packages: ['gorm.io/gorm'] },
  // Testing
  { name: 'Jest', category: 'testing', packages: ['jest'] },
  { name: 'Vitest', category: 'testing', packages: ['vitest'] },
  { name: 'Pytest', category: 'testing', packages: ['pytest'] },
];

// ─── Layer detection patterns ───

interface LayerPattern {
  name: string;
  pattern: string;
  regex: RegExp;
}

const LAYER_PATTERNS: LayerPattern[] = [
  { name: 'Controller/Handler', pattern: 'controller*|handler*|route*|api*', regex: /^(controller|handler|route|api)/i },
  { name: 'Service', pattern: 'service*|business*|domain*', regex: /^(service|business|domain)/i },
  { name: 'Repository/Data', pattern: 'repo*|repository*|dao*|data*|model*|entity*', regex: /^(repo|repository|dao|data|model|entity)/i },
  { name: 'Utility', pattern: 'util*|helper*|common*|shared*|lib*', regex: /^(util|helper|common|shared|lib)/i },
];

// ─── ArchitectureAnalyzer ───

export class ArchitectureAnalyzer implements AnalysisModuleInterface {
  getName(): string {
    return 'architecture';
  }

  async analyze(profile: ProjectProfile, plugins: LanguagePlugin[]): Promise<ModuleResult> {
    // 1. Collect dependencies from all plugins
    const dependencies = this.collectDependencies(profile, plugins);

    // 2. Group dependencies by category
    const dependencyGroups = this.groupDependencies(dependencies);

    // 3. Identify core frameworks
    const frameworks = this.identifyFrameworks(dependencies);

    // 4. Identify project layers from directory structure
    const layers = this.identifyLayers(profile.projectPath);

    // 5. Generate module dependency graph if sub-modules exist
    const moduleDependencyGraph = this.buildModuleDependencyGraph(profile);

    const result: ArchitectureResult = {
      dependencies,
      dependencyGroups,
      layers,
      frameworks,
      moduleDependencyGraph,
    };

    return result;
  }

  /**
   * Collect dependencies from all language plugins.
   */
  private collectDependencies(profile: ProjectProfile, plugins: LanguagePlugin[]): Dependency[] {
    const allDeps: Dependency[] = [];
    const seen = new Set<string>();

    for (const plugin of plugins) {
      const deps = plugin.extractDependencies(profile.projectPath);
      for (const dep of deps) {
        const key = `${dep.group ?? ''}:${dep.name}:${dep.version}`;
        if (!seen.has(key)) {
          seen.add(key);
          allDeps.push(dep);
        }
      }
    }

    return allDeps;
  }

  /**
   * Group dependencies by their category.
   */
  private groupDependencies(dependencies: Dependency[]): Record<DependencyCategory, Dependency[]> {
    const groups: Record<DependencyCategory, Dependency[]> = {
      'web-framework': [],
      'database': [],
      'cache': [],
      'message-queue': [],
      'security': [],
      'testing': [],
      'logging': [],
      'utility': [],
      'other': [],
    };

    for (const dep of dependencies) {
      groups[dep.category].push(dep);
    }

    return groups;
  }

  /**
   * Identify core frameworks from the dependency list.
   */
  private identifyFrameworks(dependencies: Dependency[]): FrameworkInfo[] {
    const frameworks: FrameworkInfo[] = [];
    const depMap = new Map<string, Dependency>();

    for (const dep of dependencies) {
      depMap.set(dep.name.toLowerCase(), dep);
    }

    for (const fw of KNOWN_FRAMEWORKS) {
      for (const pkg of fw.packages) {
        const dep = depMap.get(pkg.toLowerCase());
        if (dep) {
          frameworks.push({
            name: fw.name,
            version: dep.version,
            category: fw.category,
            evidence: [`dependency: ${dep.name}@${dep.version}`],
          });
          break; // Only add each framework once
        }
      }
    }

    return frameworks;
  }

  /**
   * Identify project layers by scanning top-level and src-level directories.
   */
  private identifyLayers(projectPath: string): LayerInfo[] {
    const layers: LayerInfo[] = [];
    const dirsToScan = [projectPath];

    // Also scan common source directories
    for (const sub of ['src', 'lib', 'app', 'pkg', 'internal']) {
      const subPath = path.join(projectPath, sub);
      if (this.isDir(subPath)) {
        dirsToScan.push(subPath);
      }
    }

    for (const lp of LAYER_PATTERNS) {
      const matchedFiles: string[] = [];
      const matchedClasses: string[] = [];

      for (const dir of dirsToScan) {
        this.findMatchingDirs(dir, projectPath, lp.regex, matchedFiles, matchedClasses);
      }

      if (matchedFiles.length > 0) {
        layers.push({
          name: lp.name,
          pattern: lp.pattern,
          classes: matchedClasses,
          files: matchedFiles,
        });
      }
    }

    return layers;
  }

  /**
   * Find directories matching a layer pattern and collect their files.
   */
  private findMatchingDirs(
    searchDir: string,
    projectRoot: string,
    regex: RegExp,
    files: string[],
    classes: string[],
  ): void {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(searchDir, { withFileTypes: true });
    } catch {
      return;
    }

    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      if (IGNORED_DIRS.has(entry.name)) continue;

      if (regex.test(entry.name)) {
        const dirPath = path.join(searchDir, entry.name);
        const dirFiles = this.listFilesRecursive(dirPath, projectRoot);
        files.push(...dirFiles);
        // Use file basenames (without extension) as class names
        for (const f of dirFiles) {
          const base = path.basename(f, path.extname(f));
          if (base && !base.startsWith('.')) {
            classes.push(base);
          }
        }
      }
    }
  }

  /**
   * Recursively list all files in a directory, returning relative paths.
   */
  private listFilesRecursive(dir: string, projectRoot: string): string[] {
    const result: string[] = [];
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return result;
    }

    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        if (!IGNORED_DIRS.has(entry.name)) {
          result.push(...this.listFilesRecursive(fullPath, projectRoot));
        }
      } else if (entry.isFile()) {
        result.push(path.relative(projectRoot, fullPath));
      }
    }

    return result;
  }

  /**
   * Build a module dependency graph from sub-modules in the profile.
   * Returns undefined if there are fewer than 2 sub-modules.
   */
  buildModuleDependencyGraph(profile: ProjectProfile): MermaidGraph | undefined {
    const modules = profile.modules;
    if (!modules || modules.length < 2) {
      return undefined;
    }

    const nodes: string[] = modules.map((m) => m.name);
    const edges: { from: string; to: string; label?: string }[] = [];

    for (const mod of modules) {
      const deps = this.detectModuleDependencies(mod, modules, profile.projectPath);
      for (const d of deps) {
        edges.push(d);
      }
    }

    // Build Mermaid syntax
    const lines: string[] = ['graph TD'];
    for (const node of nodes) {
      const safeId = this.sanitizeMermaidId(node);
      lines.push(`  ${safeId}["${node}"]`);
    }
    for (const edge of edges) {
      const fromId = this.sanitizeMermaidId(edge.from);
      const toId = this.sanitizeMermaidId(edge.to);
      if (edge.label) {
        lines.push(`  ${fromId} -->|"${edge.label}"| ${toId}`);
      } else {
        lines.push(`  ${fromId} --> ${toId}`);
      }
    }

    return {
      syntax: lines.join('\n'),
      nodes,
      edges,
    };
  }

  /**
   * Detect dependencies of a single sub-module on other sub-modules.
   */
  private detectModuleDependencies(
    mod: SubModule,
    allModules: SubModule[],
    projectRoot: string,
  ): { from: string; to: string; label?: string }[] {
    const edges: { from: string; to: string; label?: string }[] = [];
    const otherModules = allModules.filter((m) => m.name !== mod.name);

    if (otherModules.length === 0) return edges;

    const modAbsPath = path.isAbsolute(mod.path) ? mod.path : path.join(projectRoot, mod.path);

    if (mod.buildTool === 'npm' || mod.buildTool === 'yarn' || mod.buildTool === 'pnpm') {
      edges.push(...this.detectNpmDependencies(mod, otherModules, modAbsPath));
    } else if (mod.buildTool === 'go-mod') {
      edges.push(...this.detectGoDependencies(mod, otherModules, modAbsPath));
    } else if (mod.buildTool === 'pip' || mod.buildTool === 'poetry') {
      edges.push(...this.detectPythonDependencies(mod, otherModules, modAbsPath));
    }

    return edges;
  }

  /**
   * Detect npm sub-module dependencies by reading package.json.
   */
  private detectNpmDependencies(
    mod: SubModule,
    otherModules: SubModule[],
    modAbsPath: string,
  ): { from: string; to: string; label?: string }[] {
    const edges: { from: string; to: string; label?: string }[] = [];
    const pkgPath = path.join(modAbsPath, 'package.json');

    let pkgJson: Record<string, unknown>;
    try {
      pkgJson = JSON.parse(fs.readFileSync(pkgPath, 'utf-8'));
    } catch {
      return edges;
    }

    const depSections: { section: string; label: string }[] = [
      { section: 'dependencies', label: 'compile' },
      { section: 'devDependencies', label: 'test' },
      { section: 'peerDependencies', label: 'runtime' },
    ];

    for (const { section, label } of depSections) {
      const deps = pkgJson[section] as Record<string, string> | undefined;
      if (!deps) continue;
      const depNames = Object.keys(deps);
      for (const other of otherModules) {
        if (depNames.includes(other.name)) {
          edges.push({ from: mod.name, to: other.name, label });
        }
      }
    }

    return edges;
  }

  /**
   * Detect Go sub-module dependencies by reading go.mod.
   */
  private detectGoDependencies(
    mod: SubModule,
    otherModules: SubModule[],
    modAbsPath: string,
  ): { from: string; to: string; label?: string }[] {
    const edges: { from: string; to: string; label?: string }[] = [];
    const goModPath = path.join(modAbsPath, 'go.mod');

    let content: string;
    try {
      content = fs.readFileSync(goModPath, 'utf-8');
    } catch {
      return edges;
    }

    for (const other of otherModules) {
      // Check if go.mod requires the other module (by name or path)
      if (content.includes(other.name)) {
        edges.push({ from: mod.name, to: other.name, label: 'compile' });
      }
    }

    return edges;
  }

  /**
   * Detect Python sub-module dependencies by scanning .py files for imports.
   */
  private detectPythonDependencies(
    mod: SubModule,
    otherModules: SubModule[],
    modAbsPath: string,
  ): { from: string; to: string; label?: string }[] {
    const edges: { from: string; to: string; label?: string }[] = [];
    const pyFiles = this.listFilesRecursive(modAbsPath, modAbsPath).filter((f) => f.endsWith('.py'));

    // Collect all import lines from all .py files
    const importLines: string[] = [];
    for (const relFile of pyFiles) {
      const absFile = path.join(modAbsPath, relFile);
      try {
        const content = fs.readFileSync(absFile, 'utf-8');
        for (const line of content.split('\n')) {
          const trimmed = line.trim();
          if (trimmed.startsWith('import ') || trimmed.startsWith('from ')) {
            importLines.push(trimmed);
          }
        }
      } catch {
        // skip unreadable files
      }
    }

    const found = new Set<string>();
    for (const other of otherModules) {
      if (found.has(other.name)) continue;
      for (const line of importLines) {
        if (line.includes(other.name)) {
          edges.push({ from: mod.name, to: other.name, label: 'runtime' });
          found.add(other.name);
          break;
        }
      }
    }

    return edges;
  }

  /**
   * Sanitize a module name for use as a Mermaid node ID.
   */
  private sanitizeMermaidId(name: string): string {
    return name.replace(/[^a-zA-Z0-9_]/g, '_');
  }

  private isDir(p: string): boolean {
    try {
      return fs.statSync(p).isDirectory();
    } catch {
      return false;
    }
  }
}

const IGNORED_DIRS = new Set([
  'node_modules', '.git', 'dist', 'build', 'target',
  '.gradle', '__pycache__', '.venv', 'venv', '.idea', '.vscode',
]);
